add docker_swarm_service module

dariko opened the pull request:
##### ISSUE TYPE
 - New Module Pull Request

##### COMPONENT NAME
swarm_service

##### ANSIBLE VERSION
```
ansible 2.3.0 (devel e19c2f6a6d)
```

##### SUMMARY
This PR introduces a new module: swarm_service
this module accepts parameters mapping those of the ``docker service (create|update)`` command.
Service updates will use swarm facilities, requiring service rebuild (remove -> create) only if strictly necessary (service network updating, as per https://github.com/docker/docker/issues/25876). Other modification will be run 'live'

~~This module requires a docker-py version more recent than commit docker/docker-py@3ac73a2, which corrects some parameters.~~
This module requires a version of `docker-py` (now `docker)` >= 2.0

Partially fixes #4607

Commit:
add docker_swarm_service module

dariko commented:
Migrated from https://github.com/ansible/ansible-modules-core/pull/5547

dariko commented:
replaces #18962, lost to repomerge

kkarimi commented:
+1

wilkko commented:
Pulling images for services from private registries seems to work fine if I just include this to my playbook and apply it to each manager and worker node:

`- name: Login into private Docker registry
  docker_login:
    registry: private.registry.com
    username: username
    password: password
  become_user: "{{ ansible_ssh_user }}"`

I don't know why `docker service create/update` have this flag for:

`--with-registry-auth             Send registry authentication details to swarm agents
`

dariko commented:
@wilkko  Thank you :)

wilkko commented:
Test the port publishing with this module a run into a problem.

If I publish a port 443 like this
```
      docker_swarm_service:
        - name: 'proxy'
          state: 'present'
          image: 'private.registry.com/proxy:latest'
          replicas: 1
          publish:
            - { published_port: 443, target_port: 443, protocol: tcp }
```
After inspecting the service the port is not visible:
```
docker service inspect proxy
[
    {
        "ID": "8uvgudjfq44z8sltip0d8oqee",
        "Version": {
            "Index": 26913
        },
        "CreatedAt": "2016-12-23T08:54:34.30513299Z",
        "UpdatedAt": "2016-12-23T09:00:49.60919381Z",
        "Spec": {
            "Name": "proxy",
            "TaskTemplate": {
                "ContainerSpec": {
                    "Image": "private.registy.com/proxy:latest",
                    "Labels": {
                        "last_deployed": "2016-12-23T09:00:20Z"
                    },
                    "User": "root"
                },
                "Resources": {
                    "Limits": {},
                    "Reservations": {}
                },
                "RestartPolicy": {
                    "Condition": "any",
                    "Delay": 0,
                    "MaxAttempts": 5,
                    "Window": 30
                }
            },
            "Mode": {
                "Replicated": {
                    "Replicas": 1
                }
            },
            "Networks": [
                {
                    "Target": "7g09ttib0je2ez7u4ymv3f4q9"
                }
            ]
        },
        "Endpoint": {
            "Spec": {},
            "VirtualIPs": [
                {
                    "NetworkID": "7g09ttib0je2ez7u4ymv3f4q9",
                    "Addr": "10.0.0.2/24"
                }
            ]
        },
        "UpdateStatus": {
            "State": "completed",
            "StartedAt": "2016-12-23T09:00:46.841438478Z",
            "CompletedAt": "2016-12-23T09:00:49.609183487Z",
            "Message": "update completed"
        }
    }
]
```
If I add the port manually:
```
docker service update proxy --publish-add 443:443
```

The inspection of the service shows the port:
```
docker service inspect proxy
[
    {
        "ID": "8uvgudjfq44z8sltip0d8oqee",
        "Version": {
            "Index": 27004
        },
        "CreatedAt": "2016-12-23T08:54:34.30513299Z",
        "UpdatedAt": "2016-12-23T09:05:39.788314945Z",
        "Spec": {
            "Name": "proxy",
            "TaskTemplate": {
                "ContainerSpec": {
                    "Image": "private.registy.com/proxy:latest",
                    "Labels": {
                        "last_deployed": "2016-12-23T09:00:20Z"
                    },
                    "User": "root"
                },
                "Resources": {
                    "Limits": {},
                    "Reservations": {}
                },
                "RestartPolicy": {
                    "Condition": "any",
                    "Delay": 0,
                    "MaxAttempts": 5,
                    "Window": 30
                }
            },
            "Mode": {
                "Replicated": {
                    "Replicas": 1
                }
            },
            "Networks": [
                {
                    "Target": "7g09ttib0je2ez7u4ymv3f4q9"
                }
            ],
            "EndpointSpec": {
                "Mode": "vip",
                "Ports": [
                    {
                        "Protocol": "tcp",
                        "TargetPort": 443,
                        "PublishedPort": 443
                    }
                ]
            }
        },
        "Endpoint": {
            "Spec": {
                "Mode": "vip",
                "Ports": [
                    {
                        "Protocol": "tcp",
                        "TargetPort": 443,
                        "PublishedPort": 443
                    }
                ]
            },
            "Ports": [
                {
                    "Protocol": "tcp",
                    "TargetPort": 443,
                    "PublishedPort": 443
                }
            ],
            "VirtualIPs": [
                {
                    "NetworkID": "7g09ttib0je2ez7u4ymv3f4q9",
                    "Addr": "10.0.0.2/24"
                },
                {
                    "NetworkID": "an9c0l2ory2znwlu69hwf0gxx",
                    "Addr": "10.255.0.8/16"
                }
            ]
        },
        "UpdateStatus": {
            "State": "completed",
            "StartedAt": "2016-12-23T09:04:38.853748515Z",
            "CompletedAt": "2016-12-23T09:04:40.342660151Z",
            "Message": "update completed"
        }
    }
]
```

dariko commented:
@wilkko I can't reproduce your problem; can you tell me the versions of docker and docker-py you're using?

This module requires some commits to docker-py not released in the 1.x series (I'm using commit ef76c8552c7013ed10dc2dedb185ddc392685439 installed via `pip install git+https://github.com/docker/docker-py.git@ef76c8552c7013ed10dc2dedb185ddc392685439#egg=docker-py`).

I've submitted a pull request ( #19633 ) which adds compatibility for the ansible docker modules with the python docker library >=2.0.0, which includes the requeired changes.

wilkko commented:
I am running 1.12.3 (CoreOS beta) and docker-py 1.10.6. Everything else so far had worked beautifully than the published ports.

dariko commented:
Unfortunately in the docker-py commit I pointed out (https://github.com/docker/docker-py/commit/ef76c8552c7013ed10dc2dedb185ddc392685439) there are some fixes to the service creation which are needed for this module to work correctly.

Until/uness #19633 gets merged, the workaround is to upgrade docker-py on the docker hosts to the appropriate git commit. You can install it via this task:
```
- pip:
    name: git+https://github.com/docker/docker-py.git@ef76c8552c7013ed10dc2dedb185ddc392685439#egg=docker-py
    state: latest
```

wilkko commented:
OK. I'll try with that pip version. Didn't know you can install specific git revisions with pip...

dariko commented:
Let me now how it works out

wilkko commented:
Work now, thanks. Do you think the docker-py fix will be released in 1.x series?

dariko commented:
> Work now, thanks.

Good :)

> Do you think the docker-py fix will be released in 1.x series?

I don't know (I'm not involved), but the project has no longer a 1.11 milestone (only a 2.1 milestone is active).. so maybe not.

dariko committed:
reformat respecting pep8

dariko commented:
I rebased this pr to e19c2f6a6dcb7013fd1eb176befbc21ee9c2d80e, and updated the code formatting to pass the new pep8 sanity check.

dariko committed:
add dns-dns_search-dns_options parameters, client API version detection

dariko committed:
add hostname,tty parameters

dariko committed:
add support for publishmode in publish hash

dariko committed:
add ANSIBLE_METADATA

dariko commented:
I added support for new features in docker 1.25 remote API (docker >=1.13)

This module now support these additional parameters
- dns
- dns_options
- dns_search
- hostname
- tty
that mirror the respective `docker service <create-update>` parameters.

The `publish` parameter now supports a `mode` key which if configured to `host` [bypasses the routing mesh when publishing ports,](https://github.com/docker/docker.github.io/blob/master/engine/swarm/services.md#publish-ports) mirroring the `PublishMode` attribute of `docker service`'s publish parameter expressed in the new format.

These new parameters (as specified in `DOCUMENTATION`) will work only if the module is connecting to a docker process whose API version is >= 1.25.
If they are used when connecting to an older service the module fails with a message explaining the incompatibility.

christopherobin commented on lib/ansible/modules/cloud/docker/docker_swarm_service.py:
```diff
@@ -0,0 +1,902 @@
+        task_template=types.TaskTemplate(
+            container_spec  = cspec,
+            restart_policy  = restart_policy,
+            placement       = self.constraints,
```
The `constraints` argument does not work on API 1.26 since the `placement` object in the API is not a list of strings, but an object with a single key `constraints` that takes an array of strings. This can be fixed by changing this line to

```python
            placement       = { "constraints": self.constraints },
```

See https://docs.docker.com/engine/api/v1.26/#operation/ServiceCreate for more details

dariko replied:
The conversion you suggest is handled in [the docker python library](https://github.com/docker/docker-py/blob/master/docker/types/services.py#L35), and checking the older versions docs it looks like the `placement` parameter has not changed between 1.24, 1.25 and 1.26.

Can you show me the module call which gave you problems (and the error you've received)?

I tested this parameter on docker 1.12.6, 1.30.0, and 1.30.1 with api_version 1.24, 1.25 and 1.26 and got no problems; this is the syntax I used:
```
docker_swarm_service:
    name: myservice
    image: "nginx:1.10.2-alpine@sha256:56766e6df28bb0c572dca24acbe5639ddf866d931c105cb33410d52624c27682"
    constraints:
    - "node.labels.type==ok"
```

christopherobin replied:
My bad another role was installing `docker-py` (which is `1.10.6` instead of `2.x`) which resulted in some weird behaviour on my box, sorry.

Feel free to delete those comments.

Might still be useful to emit a deprecation warning when `docker.__version__` is not at least 2.x since some older roles and most distros are still installing the older version of the library.

dariko replied:
You're right about the deprecation (in fact this module needed since its inception a specific commit of docker-py which has become 2.0).
I think I'm gonna use a failure instead of a warning: the module behaviour when using docker-py (<2.0.0) is just inconsistent.

dariko committed:
fail if docker-python version <= 2.0.0

dariko committed:
reformat EXAMPLES

dariko commented:
Rebased to b2bd75a408f929fe03e2817c6bd23d831bce7607, reformatted EXAMPLES (was failing new tests).

The module now fails fail if the `version(docker-py)>=2.0.0` requirement is not sadisfied

JanKoppe commented:
I've been testing this module successfully. I've noticed that there is no option to configure secrets for services, although according to https://docs.docker.com/engine/reference/commandline/service_create/#create-a-service-with-secrets `service create` supports the `--secrets` flag. Do you intend to support those?

JanKoppe commented:
Would love to see the `update --force` option as well, to restart a service. See https://github.com/docker/docker/pull/27596

dariko commented:
@JanKoppe 
sorry for the late reply. I completely missed the `secret` part from the changelog, I'm gonna implement it in the next couple of days, probably as a list of secret names.
There will still be need for a  docker_secret module, but this is outside of this PR scope.

About the `--force` option, I think I'm gonna implement it as another value (ie: `restarted`) for the `state` parameter, which will be equivalent to 'present' and forcefully trigger an update via the `ForceUpdate` API parameter if the service is already present.

PS: this module uses the ansible docker_module functionality as per https://github.com/ansible/ansible-modules-core/pull/5547#issuecomment-259929195. The docker_module as of now does not instantiate a docker-py high-level `Client`, but a low-level `ApiClient`

edavism commented:
+1

JamesGuthrie commented:
I like the idea of this module, but I think that the API as implemented is not very practical. It would be much simpler to use if one could practically copy-paste service definitions from a swarm-stack-ready docker-compose file into a docker_swarm_service definition.

For instance currently there are four keys related to restart policies: `restart_policy`, `restart_policy_attempts`, `restart_policy_delay`, `restart_policy_window`. In docker-compose syntax these would be declared as follows:

```
deploy:
    restart_policy:
        condition: on-failure
        delay: 5s
        max_attempts: 3
        window: 120s
```

I have taken a look at moving towards a more docker-compose inspired syntax and it shouldn't be too hard to get pretty close to parity.

dariko commented:
@JamesGuthrie 
when i implemented the restart_policy was trying to adhere ansible's module parameters validation, but I think you're right: a the swarm service API exposes too many options for them to be flattened. A compose-like syntax would be a more usable (and reusable) approach

MurgaNikolay commented:
Hi guys.
I try to setup servicea via this module, in the same time update existed services (created manual via cli). But I have error for some services:

      { "changed": false, "failed": true, "msg": "Error looking for service named reat-api: list index out of range"}

This error occurs for services created manually.
I found out that manual created services contain a network name in `raw_network_data['Target']` instead network ID:
    
    ...
    "Networks": [
        {
            "Target": "api_network"
        }
    ]
    ...

I fix than error modified 

https://github.com/dariko/ansible/blob/a09b6bb2e6466c523bf2f77df54f816f916874e1/lib/ansible/modules/cloud/docker/docker_swarm_service.py#L733
 
    for raw_network_data in raw_data['Spec'].get('Networks',[]):
      network_name=[network_name_id['name'] for network_name_id in networks_names_ids if network_name_id['id']==raw_network_data['Target']]
      if len(network_name) == 0:
        ds.networks.append(raw_network_data['Target'])
      else:
        ds.networks.append(network_name[0])

Docker version 1.13.0, build 49bf474

dariko committed:
correct service spec.network parsing, consider name from Target

dariko commented:
@MurgaNikolay: thanks, I've commited you changes

dariko committed:
pep8 style corrections

dariko commented:
I can't open the details of the shippable run (404 when clicking on 'Details'): how can I see which check is failed?

dariko committed:
correct typos in comparisons

dariko committed:
correct documentation

dariko committed:
trailing whitespace in documentation

dariko committed:
correct ansible metadata, documentation requirements

dariko committed:
explode imports from module_utils.docker_common

dariko commented:
@JamesGuthrie 
following your suggestion I've gone ahead and built an (embrional) module for managing docker stacks.
It is by no way ready for submission here (it does not use ansible docker base classes, for once), but it is surviving some initial testing. If you want to take a look you can find it here: https://github.com/dariko/ansible-module-docker_stack

rikonor commented:
I'm seeing an issue when trying to create a service using this module.

My Ansible is on the branch/commit associated with this PR.
The node I am running `docker_swarm_service` on has the latest `docker` python module installed (2.3 I believe).

I am running the following:
```
---
- name: deploy dns-auto-update as a docker swarm service
  hosts: homebase
  remote_user: pi
  tasks:
    - name: deploy dns-auto-update as a docker swarm service
      docker_swarm_service:
        name: dns-auto-update
        image: registry.rikonor.xyz/dns-auto-update
        state: present
        replicas: 1
```
The service gets created but fails immediately with the following error:
```
ID                          NAME                IMAGE                                                                                                                 NODE                DESIRED STATE       CURRENT STATE           ERROR                                                                                                        PORTS
vsq72tp2voywoovqj0rcfrzmj   dns-auto-update.1   registry.rikonor.xyz/dns-auto-update:latest@sha256:c2544cdebd5b2fb8c6b1b3be01a384223c14413977525eb5744fe413af194e64   pi3                 Shutdown            Failed 16 seconds ago   "starting container failed: linux spec user: unable to find user root: no matching entries in passwd file"
```
Any ideas what might be going on?

I am able to run the service just fine by running:
```
docker service create \
  --name dns-auto-update \
  --replicas 1 \
  registry.rikonor.xyz/dns-auto-update
```
I doubt this is related but just in case: the image I'm trying to run contains a single binary in a `scratch` image so I'm not even sure there is a `passwd` file in it.

dariko commented:
@rikonor 
Can you post the output of the `docker service inspect dns-auto-update` command after creating the service both via CLI and ansible?
Was `docker service create` run as `pi`?

Ranger-X commented:
@rikonor
Try to add option: ``` user: ''``` to the ```docker_swarm_service```

rikonor commented:
@Ranger-X Good call. That indeed fixed it.

That said, would be nice not to have to do that.

Here are the different `docker service inspect X` responses. Sorry for the wall of text.

```
# Creating service via CLI
[
    {
        "ID": "sji365u5kz3wam87hq906yk9b",
        "Version": {
            "Index": 1680
        },
        "CreatedAt": "2017-05-21T14:58:53.105958918Z",
        "UpdatedAt": "2017-05-21T14:58:53.105958918Z",
        "Spec": {
            "Name": "megusta",
            "Labels": {},
            "TaskTemplate": {
                "ContainerSpec": {
                    "Image": "registry.rikonor.xyz/megusta",
                    "Env": [],
                    "DNSConfig": {}
                },
                "Resources": {
                    "Limits": {},
                    "Reservations": {}
                },
                "RestartPolicy": {
                    "Condition": "any",
                    "MaxAttempts": 0
                },
                "Placement": {},
                "ForceUpdate": 0
            },
            "Mode": {
                "Replicated": {
                    "Replicas": 1
                }
            },
            "UpdateConfig": {
                "Parallelism": 1,
                "FailureAction": "pause",
                "MaxFailureRatio": 0
            },
            "RollbackConfig": {
                "Parallelism": 1,
                "FailureAction": "pause",
                "MaxFailureRatio": 0
            },
            "EndpointSpec": {
                "Mode": "vip"
            }
        },
        "Endpoint": {
            "Spec": {}
        }
    }
]
```

```
# Creating service via Ansible
[
    {
        "ID": "mfbhq33w1ir63f9bwf0l9zeo8",
        "Version": {
            "Index": 1668
        },
        "CreatedAt": "2017-05-21T14:57:53.687185703Z",
        "UpdatedAt": "2017-05-21T14:57:53.687185703Z",
        "Spec": {
            "Name": "megusta",
            "Labels": {},
            "TaskTemplate": {
                "ContainerSpec": {
                    "Image": "registry.rikonor.xyz/megusta:latest@sha256:c0dc78a8c128d5ba8b16602706c6de25cba09ae65cb40684cd476a2efb8a7f64",
                    "Env": [],
                    "User": "root",
                    "DNSConfig": {}
                },
                "Resources": {
                    "Limits": {},
                    "Reservations": {}
                },
                "RestartPolicy": {
                    "Condition": "none",
                    "Delay": 0,
                    "MaxAttempts": 0,
                    "Window": 0
                },
                "Placement": {},
                "ForceUpdate": 0
            },
            "Mode": {
                "Replicated": {
                    "Replicas": 1
                }
            },
            "EndpointSpec": {
                "Mode": "vip"
            }
        },
        "Endpoint": {
            "Spec": {}
        }
    }
]
```

```
# Creating service via Ansible with `user: ''` under `docker_swarm_service`
[
    {
        "ID": "6jcs5ecenzbfkbj1mj2yn0b6t",
        "Version": {
            "Index": 1658
        },
        "CreatedAt": "2017-05-21T14:55:34.881815978Z",
        "UpdatedAt": "2017-05-21T14:55:34.881815978Z",
        "Spec": {
            "Name": "megusta",
            "Labels": {},
            "TaskTemplate": {
                "ContainerSpec": {
                    "Image": "registry.rikonor.xyz/megusta:latest@sha256:c0dc78a8c128d5ba8b16602706c6de25cba09ae65cb40684cd476a2efb8a7f64",
                    "Env": [],
                    "DNSConfig": {}
                },
                "Resources": {
                    "Limits": {},
                    "Reservations": {}
                },
                "RestartPolicy": {
                    "Condition": "none",
                    "Delay": 0,
                    "MaxAttempts": 0,
                    "Window": 0
                },
                "Placement": {},
                "ForceUpdate": 0
            },
            "Mode": {
                "Replicated": {
                    "Replicas": 1
                }
            },
            "EndpointSpec": {
                "Mode": "vip"
            }
        },
        "Endpoint": {
            "Spec": {}
        }
    }
]
```

Only difference I can tell is that perhaps the user should not be set to `root` by default?
Another observation is that there are a bunch of different defaults when creating via CLI, e.g. `RestartPolicy`.

dariko commented:
The problem is in the default for the parameter `user`, which right now is `root`. I will try to disable the module default value and let the API take care of it, but I'll need to explore how this will influence other situations.

rikonor commented:
Hey @dariko,

I wanted to ask about the API provided by this module. It seems to differ in language from the other modules (e.g `docker_container`).

Instead of the expected `ports` it exposes `published`.
Instead of the expected `volumes` it exposes `mounts`.
Instead of the expected `command` it exposes `args`.

dariko commented:
@rikonor The docker naming is different when talking about containers and services: the module parameters names reflect those of docker commands [`service create`](https://docs.docker.com/engine/reference/commandline/service_create/) and [`service update`](https://docs.docker.com/engine/reference/commandline/service_update/).

To have a service restart unconditionally you should set `restart-condition` to `any` and leave `restart-max-attempts` to `0`.

rikonor commented:
@dariko Ah, I see that now. Thanks!

MurgaNikolay commented:
Hi @dariko.
Could you review pull request to this module https://github.com/dariko/ansible/pull/1
Thanks.

dariko commented:
@MurgaNikolay thanks, I merged your changes

dariko commented:
@MurgaNikolay Could you resubmit your PR to the branche referenced in this PR? (https://github.com/dariko/ansible/tree/ansible-modules-core/pull/5547)
The one you referenced (https://github.com/dariko/ansible/tree/docker_swarm_service) is a stale one forked when the `ansible-modules-core` repository was being merged

soumyadipDe commented:
Which version can I expect this to be available?

phjardas commented:
Has there been any progress on this PR? We'd really like to use ansible to deploy to docker swarm.

rikonor commented:
For what it's worth, you can always fork the latest version of Ansible and manually add this file to your fork.

Ranger-X commented:
If I pass an option 'limit_memory: 8192', I get an error: {"changed": false, "failed": true, "msg": "Error parsing module parameters: global name 'human_to_bytes' is not defined"}. In which python-module is this function defined?

dariko commented:
@Ranger-X It was defined in `docker-common.py`, and got removed by fd44917561c4813735c1832fddb34596131242dd. To be honest I didn't import it when I exploded the imports from docker_common.

I can't test it now, but could you try adding the line
```
from ansible.module_utils.basic import human_to_bytes
```
after 
```
from ansible.module_utils.docker_common import docker_version
```
and test it?

Anyway I should be able to test it in a couple days

Ranger-X commented:
@dariko Thank you! It is working now after applying your suggest.

jctanner commented:
bot_status

dariko committed:
use remaining version of human_to_bytes after fd44917561c4813735c1832fddb34596131242dd

dariko committed:
update metadata_version, version_added

dariko committed:
update RETURN metadata

dariko committed:
subordinate docker_version import to docker_common.py library checks

dariko committed:
cleanup headers

dariko committed:
add boilerplate as per 1ee08c0f069eb5d3817ba7199adef4f3804056b8

dariko committed:
replace call to unicode() with ansible.module_utils._text.to_text()

dariko commented:
I rebased to devel and cleaned up the code/metadata for the new CI checks

soumyadipDe commented:
Is this up for next release?

LuisGAngel commented:
^ Is this up for next release?

ricalenil commented:
In what release is going to be this module?

mcassaniti commented:
Can I suggest having the env specified as a dictionary rather than a list? I know that this doesn't match the `docker service create` specification, but it certainly makes a lot of sense inside Ansible. Most of your other justifications for matching `docker service create|update` make perfect sense.

DimmKirr commented on lib/ansible/modules/cloud/docker/docker_swarm_service.py:
```diff
@@ -0,0 +1,905 @@
+
+    def get_service(self, name):
+        raw_data = self.client.services(filters={'name': name})
+        if len(raw_data) == 0:
```
Seems like filter can return partial matches: `foo` would match for an existing `foo-bar` service, and would fail with thinking it needs to be updated (which also seems to be deprecated. Maybe it's worth verifying with this condition and adding a workaround:
```python
if len(raw_data) == 0 or not 'Name' in raw_data[0]['Spec'] or raw_data[0]['Spec']['Name'] != name:
```

ricalenil replied:
Do a pull request.

DimmKirr replied:
There were so many migrations, I'm confused where is the original PR coming from. Is it https://github.com/dariko/ansible-modules-core/blob/swarm_service/cloud/docker/docker_swarm_service.py ?

ricalenil replied:
Yes

ricalenil replied:
Yes, it is.

dariko replied:
@kireevco sorry for the late reply. Gitub sure doesn't make it is to find the source branch of a PR :) I'm confident it's this one: https://github.com/dariko/ansible/tree/docker_swarm_service

dariko replied:
@kireevco  I was confident, and I was wrong. The source of this PR is https://github.com/dariko/ansible/tree/ansible-modules-core/pull/5547.
Sorry, but this PR have a messy and long history.

inhumantsar commented:
@dariko should this be closed in favor of https://github.com/ansible/ansible/pull/24588 ?

jwitko commented:
So I figured out the issues here (I think). 

1) When attempting to import docker modules and methods (Specifically, `Ulimit`) the try clause was failing due to changes in the `docker-py` module.  This failure caused `docker_version` to not be defined and caused the error I listed above.  I fixed this by using `LooseVersion` and the `docker` module provided `__version__`. 

2)  It looks like the docker-py `create_service()` method used to look for `endpoint_config` parameter but was changed in https://github.com/docker/docker-py/pull/1239/files#diff-c868ccedada5d47fa6217ad089d1e5d3L30 to look for `endpoint_spec`.

I have a PR ready to go but I cannot figure out for the life of me the originating branch.  I see its talked about above but those branches do not match the current state of the file in this PR at all.  @dariko Let me know if you can point me to the correct file and I'll happily submit the fixes.

dariko commented:
@jwitko Thank you, I've tried your patch and it's good
The source of this PR is https://github.com/dariko/ansible/tree/ansible-modules-core/pull/5547, I think you can request a pull from there.
If you prefer I can commit it myself.

jwitko committed:
Update docker_swarm_service.py

- Fix version checking
- Fix change in docker-py spec for endpoint_config to endpoint_spec

jwitko commented:
@dariko put a PR up here https://github.com/dariko/ansible/pull/3

dariko commented:
thank you, I've merged your PR.

dariko committed:
corrections for E307, E322, E324, E325

jctanner commented:
!waffling needs_revision

dariko committed:
correct documentation

ci error:
-----
docs/docsite/rst/modules/docker_swarm_service_module.rst:21:0: warning:
Bullet list ends without a blank line; unexpected unindent.
-----

jctanner commented:
bot_status

NikolayMurha commented:
Hi guys.

@dariko I have version of this module with some changes 
https://gist.github.com/NikolayMurha/318fd769b5cbd87e3c0862c202ca0d00#file-docker_swarm_service-py

- Add params: `secrets: []`, `configs: []`
- Rewrite `generate_docker_py_service_description` to use `dokey-py` type (https://gist.github.com/NikolayMurha/318fd769b5cbd87e3c0862c202ca0d00#file-docker_swarm_service-py-L688)
- Fix `reserver_memory`, `reserver_cpu`, `limit_cpu`, `limit_memory` 
- Fix fetch "networks" for `current_service`  For docker api > 1.25 Networks moved from `Spec` section to 'Spec.TaskTemplate' (https://github.com/docker/docker-py/pull/1804) https://gist.github.com/NikolayMurha/318fd769b5cbd87e3c0862c202ca0d00#file-docker_swarm_service-py-L851

Sorry guys, I have no time for make pull request. So If you find these changes useful, I'll be happy if they fall into the main branch.
Thanks.

tbouvet commented:
Hi guys,

Do you need some help to finish this PR ? I have tested it and it seems to be good for me.
thanks @dariko

jwitko commented:
@NikolayMurha's changes are probably worth incorporating.  I've been using them for about a month now pretty successfully.  

@dariko I can submit another PR to the main branch to incorporate them.

Commit:
Add params: secrets, configs, log_driver, log_driver_options
Rewrite generate_docker_py_service_description to use docker-py types
Fix reserver_memory, reserver_cpu, limit_cpu, limit_memory
Fix fetch "networks" for current_service For docker api > 1.25 Networks moved from Spec section to Spec.TaskTemplate

jwitko commented:
@dariko https://github.com/dariko/ansible/pull/4

dariko committed:
fix ci errors: upgraded version_added, aligned indentation

dariko commented:
Great!
Thank you both @jwitko and @NikolayMurha.

I'm fixing a couple of CI errors

dariko commented:
@jwitko 
sorry to be late...
anyway:
1.  It seems to be working: I run [this script](https://gist.github.com/dariko/c9e04f7ae55e1ee9024a8567660ad1a6) and the resulting service had the correct attributes
`````
(ansible) [root@docker-test ~]# docker service inspect service
[ CUT ]
                "Resources": {
                    "Limits": {
                        "MemoryBytes": 20971520
                    },
                    "Reservations": {
                        "MemoryBytes": 10485760
                    }
[ CUT ]
(ansible) [root@docker-test ~]#
`````
2. I think it could be useful, but I'm using docker only marginally these days so I don't know when I'll happen to dwelve again in this module. PRs are always welcome :)
   Anyway, if you can use a stack instead of a service I'd suggest you give a try to [this module](https://github.com/ansible/ansible/pull/19229#issuecomment-298054361) which delegate to the docker cli all the logic which is explicit in this one.

jwitko committed:
Add container update policy config options

This PR adds the ability for the user to specify update policy settings like delay, parallelism, etc.

dariko committed:
PEP fixes (E501, E231, E305)

dariko committed:
Document choices for update_failure_action, update_order parameters

dariko commented:
@jwitko thank you, I've merged your PR

jwitko committed:
Fix error for when swarm mode is

jwitko committed:
Add lookups for existing values of updateConfig keys

Adding correct lookups of existing UpdateConfig policy key/values in order to generate correct OK/CHANGED response.

jwitko commented:
@dariko Have another one for you, discovered some bugs with module return status due to not looking up existing values properly for comparison.  Please see https://github.com/dariko/ansible/pull/6 , I believe after that we can push for review.

dariko commented:
@jwitko 
thank you:)
As I wrote on dariko#6 comments, I addeded you as a collaborator for the PR repo: I hope this will streamline your contributions.

jwitko committed:
Adding myself as co-author, fixing whitespace issues

jwitko commented on lib/ansible/modules/cloud/docker/docker_swarm_service.py:
```diff
@@ -12,7 +12,7 @@
 ---
 module: docker_swarm_service
-author: "Dario Zanzico (@dariko)"
+author: "Dario Zanzico (@dariko), Jason Witkowski (@jwitko)"
```
@dariko Let me know if you have an issue with this.

dariko replied:
@jwitko No issue at all, looks good to me

mcassaniti commented:
Can we get a state like update-force? I know it's a bit of an odd question, but I do have times where running `docker service update --force <service name>` is required.

mcassaniti commented:
I'm also getting an error `no networks named smtp`, but I can clearly see a swarm network named smtp. I can also create my service by hand with that network name, so I assume it is all correct.

jwitko commented:
@mcassaniti I use the module every day, with networks defined and also allowing the default behavior in places.  I have had no issues at all.  Can you please post some more information regarding your issue?  Maybe things like the ansible code executing the call, the commit id of the module you're using in your testing, some of the debug output of your swarm cluster regarding the networking that is having issues, etc.

jwitko committed:
add support for force_update flag

jwitko commented:
Added support for `docker service update --force` flag

jwitko committed:
fixing whitespace after , for pip8

ryansb commented:
bot_status

jwitko commented:
@cove @felixfontein @kassiansun @softzilla 
Went to a core #ansible-meeting on irc and it is requested that I obtain a review from one or more of you.  Would anyone be so kind?

felixfontein commented:
@jwitko I should be able to take a look at it this weekend.

jjjbushjjj commented:
Just want to say, Thank you guys for this module! Now I can rid of my ugly (shell,command) constuctions in the palybooks.

mcassaniti commented:
Thanks for update_force, and for the module in general.

After taking a look at the code, I can see that the network list comes from `ansible.module_utils.docker_common.AnsibleDockerClient`. It's probably got to do with the version of the python docker module on the target node.
